beard/clustering/preclustering.py

```diff
@@ -0,0 +1,392 @@
+        self._single_names_variants = 1
+        self._name = tokens[0][-1]
+
+    def add_signature(self, tokens):
+        """Add a signature to the precluster.
+
+        Parameters
+        ----------
+        :param tokens: tuple
+            Tokens created from the author name in form of
+            ((surname tokens), (first name tokens)).
+        """
+        if tokens[0] in self._content:
+            if tokens[1] in self._content[tokens[0]]:
+                self._content[tokens[0]][tokens[1]] += 1
+            else:
+                self._content[tokens[0]][tokens[1]] = 1
+        else:
+            self._content[tokens[0]] = {tokens[1]: 1}
+
+        if len(tokens[0]) == 1:
+            self._single_names_variants += 1
+
+    def compare_tokens_from_back(self, tokenized_prefix, last_name):
+        """Match a part of the surname with given names in precluster.
+
+        For example, ``Sanchez-Gomez, Juan`` can appear on a signature as
+        ``Gomez, Juan Sanchez``. This function checks if there is a match
+        between a signature like ``Sanchez-Gomez, Juan``, and names inside the
+        precluster.
+
+        Parameters
+        ----------
+        :param tokenized_prefix: tuple
+            Tokens which represent all given names with few first surnames
+            added. In form of a tuple of strings
+        :param last_name: tuple
+            Tokens, usually one, representing last surname(s) of the author.
+
+        Raises
+        ------
+        :raises: KeyError
+            When the last name is not included in the cluster
+
+        Returns
+        -------
+        :returns: boolean
+            Information whether cluster contains this author if few last names
+            are treated as first names.
+        """
+        if last_name in self._content:
+            for first_names in six.iterkeys(self._content[last_name]):
+                first_names_left = len(first_names)
+                for reversed_index, name in \
+                        enumerate(reversed(tokenized_prefix)):
+                    if first_names_left == 0:
+                        return True
+                    elif first_names[-(reversed_index + 1)] != name:
+                        break
+                    first_names_left -= 1
+                    if reversed_index == len(tokenized_prefix) - 1:
+                        return True
+            return False
+        self._raise_keyerror(last_name)
+
+    def contains(self, tokens):
+        """Check if there is at least one signature with given surnames.
+
+        Parameters
+        ----------
+        :param tokens: tuple
+            Tokens which represent all surnames. Tuple of strings
+
+        Returns
+        -------
+        :returns: boolean
+            True if there is at least one sinature with given surnames.
+        """
+        return tokens in self._content
+
+    def initials_score(self, new_first_names, last_name):
+        """Count matches among the initials.
```

Comment:
I think this also compares given names not necessary initials. Could we rename the function to something like 'given_names_matching_score'?